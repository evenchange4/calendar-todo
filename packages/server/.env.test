PORT=4000
DEBUG=server

GOOGLE_CLIENT_ID=GOOGLE_CLIENT_ID
GOOGLE_CLIENT_SECRET=GOOGLE_CLIENT_SECRET
GOOGLE_REDIRECT_URL=GOOGLE_REDIRECT_URL
