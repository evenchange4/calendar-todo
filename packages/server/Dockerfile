FROM mhart/alpine-node:10.24.1 AS builder
WORKDIR /app
COPY . .
RUN yarn install --pure-lockfile
RUN yarn run build
RUN yarn run pkg

FROM alpine:3.8
RUN apk update && \
  apk add --no-cache curl libstdc++ libgcc ca-certificates && \
  rm -rf /var/cache/apk/*
WORKDIR /app
COPY --from=builder /app/pkg .
ENV NODE_ENV=production
ENV DEBUG=server
ENV PORT=4000
HEALTHCHECK --start-period=30s --interval=30s --timeout=5s \
  CMD curl --fail http://localhost:${PORT}/health || exit 1
CMD ./server

