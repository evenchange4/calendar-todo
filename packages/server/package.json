{
  "private": true,
  "name": "server",
  "version": "0.0.0",
  "license": "MIT",
  "bin": "lib/index.js",
  "pkg": {
    "assets": [
      ".env.example"
    ],
    "scripts": [
      "lib/**/*.js"
    ]
  },
  "scripts-info": {
    "dev": "[Development] Run a livereload dev server.",
    "build": "[Production] Compile source code to CJS with babel.",
    "start": "[Production] Run a production server.",
    "pkg": "[Production] Package into an binary."
  },
  "scripts": {
    "dev": "nodemon src/index.js --watch src --exec babel-node",
    "build": "rm -rf lib && NODE_ENV='production' babel src -d lib --ignore '*/__tests__,*/__mock-data__'",
    "start": "node ./lib/index.js",
    "pkg": "pkg . --targets node10-alpine-x64 --out-path pkg",
    "test": "NODE_ENV='test' TZ='UTC' jest --coverage",
    "test:watch": "npm run test -- --watch"
  },
  "dependencies": {
    "apollo-server": "2.0.0",
    "apollo-server-express": "2.0.0",
    "chalk": "2.4.1",
    "compression": "1.7.3",
    "cors": "2.8.4",
    "debug": "3.1.0",
    "dotenv-safe": "6.0.0",
    "express": "4.16.3",
    "fs-extra": "7.0.0",
    "googleapis": "32.0.0",
    "graphql": "0.13.2",
    "helmet": "3.13.0",
    "morgan": "1.9.0"
  },
  "devDependencies": {
    "@babel/cli": "7.0.0-beta.56",
    "@babel/core": "7.0.0-beta.56",
    "@babel/node": "7.0.0-beta.56",
    "@babel/plugin-proposal-object-rest-spread": "7.0.0-beta.56",
    "@babel/preset-env": "7.0.0-beta.56",
    "@babel/preset-flow": "7.0.0-beta.56",
    "babel-core": "7.0.0-bridge.0",
    "babel-jest": "23.4.2",
    "babel-plugin-add-module-exports": "0.3.3",
    "fkill": "5.3.0",
    "flow-bin": "0.78.0",
    "get-port": "4.0.0",
    "graphql-request": "1.8.0",
    "jest": "23.4.2",
    "nodemon": "1.18.3",
    "pkg": "4.3.3"
  },
  "jest": {
    "testEnvironment": "node",
    "setupTestFrameworkScriptFile": "<rootDir>/setupTests.js",
    "collectCoverageFrom": [
      "src/**/*.{js,jsx}",
      "!src/**/*.test.{js,jsx}"
    ],
    "testPathIgnorePatterns": [
      "<rootDir>/node_modules/",
      "<rootDir>/lib/"
    ],
    "resetMocks": true,
    "resetModules": true
  },
  "babel": {
    "presets": [
      [
        "@babel/preset-env",
        {
          "targets": {
            "node": "current"
          }
        }
      ],
      "@babel/preset-flow"
    ],
    "plugins": [
      "babel-plugin-add-module-exports",
      "@babel/plugin-proposal-object-rest-spread"
    ]
  }
}
