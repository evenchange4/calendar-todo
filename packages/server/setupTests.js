// @flow

require('dotenv-safe').load({ path: '.env.test' });

jest.mock('./src/utils/getOauthClient', () => () => ({}));
