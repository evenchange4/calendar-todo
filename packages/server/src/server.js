/* eslint import/first: 0 */
// @flow

import './utils/dotenv';
import express from 'express';
import { ApolloServer } from 'apollo-server-express';
import { google } from 'googleapis';
import compression from 'compression';
import helmet from 'helmet';
import morgan from 'morgan';
import cors from 'cors';
import typeDefs from './modules/typeDefs';
import resolvers from './modules/resolvers';
import log from './utils/log';
import getOauthClient from './utils/getOauthClient';

const server = async (port: string) => {
  const auth = await getOauthClient();
  const calendar = google.calendar({ version: 'v3', auth });
  const apolloServer = new ApolloServer({
    typeDefs,
    resolvers,
    context: { calendar },
  });
  const app = express();
  app.use(compression({ threshold: 0 }));
  app.use(helmet());
  app.use(morgan('combined'));
  app.use(cors());
  (app: any).get('/health', (req, res) => res.send('200 OK')); // Note: For HEALTHCHECK

  apolloServer.applyMiddleware({ app });

  app.listen({ port }, () =>
    log(`ðŸš€ Server ready at http://localhost:4000${server.graphqlPath}`),
  );
};

export default server;
