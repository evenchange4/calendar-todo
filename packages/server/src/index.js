/* eslint import/first: 0 */
// @flow

import './utils/dotenv';
import server from './server';

const { PORT }: any = process.env;

server(PORT);
