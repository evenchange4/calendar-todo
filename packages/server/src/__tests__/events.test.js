// @flow
import getPort from 'get-port';
import { request } from 'graphql-request';
import fkill from 'fkill';
import server from '../server';
import mockData from '../__mock-data__/events.json';

jest.mock('googleapis', () => ({
  google: {
    calendar: () => ({
      events: {
        list: async () => mockData,
      },
    }),
  },
}));

describe('rest modules', () => {
  let port;

  beforeEach(async () => {
    port = await getPort();
    await server(port);
  });

  afterAll(async () => {
    await fkill(`:${port}`);
  });

  it('query Events', async () => {
    try {
      const results = await request(
        `http://localhost:${port}/graphql`,
        `query events {
          events {
            id
            summary
            location
          }
      }
      `,
      );
      expect(results).toMatchSnapshot();
    } catch (error) {
      expect(error).toBeUndefined();
    }
  });
});
