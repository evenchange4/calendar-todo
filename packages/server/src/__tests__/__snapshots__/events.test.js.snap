// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rest modules query Events 1`] = `
Object {
  "events": Array [
    Object {
      "id": "sfodsvfr8btnaciku1ktg464m0",
      "location": "Taipei",
      "summary": "AWS Online Tech Talks: On-Demand Library",
    },
    Object {
      "id": "0kv1j0ol9d02ihca1p8qcv5trc",
      "location": "taipei",
      "summary": "Workshop 1: AWS 101 - Introduction to AWS Cloud Computing",
    },
  ],
}
`;
